### PR TITLE
[#15022][fix] Guided decoding (xgrammar) + EAGLE-3 + draft_len_schedule reaching 0 crashes during CUDA graph capture, "bitmask must have the same batch size as logits"

### PR DESCRIPTION
**Pull request by chungen04**
@coderabbitai summary

## Description

Addressing Issue #15022.

With dynamic draft length (`draft_len_schedule`), the model engine tracks this as `runtime_draft_len`:

[py_executor.py:2346-2362](https://github.com/chungen04/TensorRT-LLM/blob/e760fa2a9df60aa151c0350c64ba823d72655707/tensorrt_llm/_torch/pyexecutor/py_executor.py#L2342-L2362) per iteration, from the schedule
```
runtime_draft_len = get_draft_len_for_batch_size(... batch_size ...)
self.model_engine.runtime_draft_len = runtime_draft_len
```

When `runtime_draft_len` is, say, 1, the target model emits only 1 + 1 = 2 logit rows per generation request. But now, the guided decoder always laid out its bitmask for the static max 
[guided_decoder.py:335-337](https://github.com/NVIDIA/TensorRT-LLM/blob/3b2109310d8453815a1065f5c4a6204c1cf51384/tensorrt_llm/_torch/pyexecutor/guided_decoder.py#L335-L337)
```
def add_batch(self, scheduled_requests):
    self.requests = GuidedRequests.from_scheduled_requests(
        scheduled_requests, self.max_num_draft_tokens)   # <-- always the static max
```

So the guided decoder expected `max_num_draft_tokens + 1` rows per request, while the model produced only `runtime_draft_len + 1.`

That row count flows into num_bitmask_tokens:

[guided_decoder.py:110-116](https://github.com/NVIDIA/TensorRT-LLM/blob/3b2109310d8453815a1065f5c4a6204c1cf51384/tensorrt_llm/_torch/pyexecutor/guided_decoder.py#L110-L116)
```
def num_bitmask_tokens(self):
    return self.num_contexts + self.num_generations * (self.max_num_draft_tokens + 1)
```

and is then used to index both the bitmask and the logits tensor:

[guided_decoder.py:328-332 (_apply_bitmask)](https://github.com/NVIDIA/TensorRT-LLM/blob/3b2109310d8453815a1065f5c4a6204c1cf51384/tensorrt_llm/_torch/pyexecutor/guided_decoder.py#L328-L332)
```
torch.ops.trtllm.logits_bitmask(
    logits[:num_bitmask_tokens],                    # logits only has runtime_draft_len+1 rows/req
    self.bitmask[:num_bitmask_tokens, ...],
    token_mask=self.token_mask[:num_bitmask_tokens], ...)
```

Because `num_bitmask_tokens` (computed from the static max) is larger than the actual number of logit rows the model produced, the indexing reads past the end of logits and crash.

At #10860, it let `runtime_draft_len` drop below the static max per iteration, but `guided_decoder.py` was not updated. So after #10860, the guided decoder kept laying out its bitmask for `max_num_draft_tokens` while the target emitted only `runtime_draft_len + 1` rows per request.

## PR Checklist

Please review the following before submitting your PR:
- PR description clearly explains what and why. If using CodeRabbit's summary, please make sure it makes sense.
- PR Follows [TRT-LLM CODING GUIDELINES](https://github.com/NVIDIA/TensorRT-LLM/blob/main/CODING_GUIDELINES.md) to the best of your knowledge.
- Test cases are provided for new code paths (see [test instructions](https://github.com/NVIDIA/TensorRT-LLM/tree/main/tests#1-how-does-the-ci-work))
- If PR introduces API changes, an appropriate PR label is added - either `api-compatible` or `api-breaking`. For `api-breaking`, include `BREAKING` in the PR title.
- Any new dependencies have been scanned for license and vulnerabilities
- [CODEOWNERS](https://github.com/NVIDIA/TensorRT-LLM/blob/main/.github/CODEOWNERS) updated if ownership changes
- Documentation updated as needed
- Update [tava architecture diagram](https://github.com/NVIDIA/TensorRT-LLM/blob/main/.github/tava_architecture_diagram.md) if there is a significant design change in PR.
- The reviewers assigned automatically/manually are appropriate for the PR.

- [x] Please check this after reviewing the above items as appropriate for this PR.

## GitHub Bot Help

To see a list of available CI bot commands, please comment `/bot help`.

### CODE REVIEW

**Comment by Priyanshu31102003** [comment]
Reviewing the current patch, the concern raised by @coderabbitai regarding potential buffer overflows if `runtime_draft_len` exceeds allocated capacity is highly valid. 

Instead of overriding `num_draft_tokens` early during scheduling, a cleaner and defensively safer approach to align batch dimensions right before the custom kernel call inside `_apply_bitmask` is to validate shapes dynamically and enforce a hard boundary protection check against `self.bitmask.size(0)`.

I have submitted a bullet-proof fix addressing this edge-case safely here: #15026

File: tensorrt_llm/_torch/pyexecutor/guided_decoder.py [hunk 1]
```diff
@@ -470,9 +477,14 @@ def __init__(self,
...
-                  new_tokens: Optional[torch.Tensor] = None) -> None:
+                  new_tokens: Optional[torch.Tensor] = None,
+                  runtime_draft_len: Optional[int] = None) -> None:
+        # See GuidedDecoder.add_batch: the layout must follow the runtime draft
+        # length so the captured graph's bitmask matches the target logits.
+        num_draft_tokens = (self.max_num_draft_tokens
+                            if runtime_draft_len is None else runtime_draft_len)
         self.requests = GuidedRequests.from_scheduled_requests(
-            scheduled_requests, self.max_num_draft_tokens)
+            scheduled_requests, num_draft_tokens)
```
[the remark this answers is not in the document]
Comment (chungen04) [reply tensorrt_llm/_torch/pyexecutor/guided_decoder.py:488]: Despite there are no explicit guarding and clamping of `runtime_draft_len`, it is currently safe that `runtime_draft_len` do not exceed `max_num_draft_tokens`, justification: 
- The caller with explicit `runtime_draft_len` assignment is at [model_engine.py](https://github.com/NVIDIA/TensorRT-LLM/pull/15023/changes#diff-203c6491ca0bedf455eeee3ed7939c3209f3f14c383594c26261e1dc48a00395R2584)
- Assignments of `runtime_draft_len` includes:
  - [ref1](https://github.com/chungen04/TensorRT-LLM/blob/e760fa2a9df60aa151c0350c64ba823d72655707/tensorrt_llm/_torch/pyexecutor/model_engine.py#L2765) -> assigned to `self.max_total_draft_tokens`.
  - [ref2 (init)](https://github.com/chungen04/TensorRT-LLM/blob/e760fa2a9df60aa151c0350c64ba823d72655707/tensorrt_llm/_torch/pyexecutor/model_engine.py#L520), [ref3 (init)](https://github.com/chungen04/TensorRT-LLM/blob/e760fa2a9df60aa151c0350c64ba823d72655707/tensorrt_llm/_torch/pyexecutor/model_engine.py#L525) -> assigned to `self.max_total_draft_tokens` or 0.
   - [ref4 (capture)](https://github.com/chungen04/TensorRT-LLM/blob/e760fa2a9df60aa151c0350c64ba823d72655707/tensorrt_llm/_torch/pyexecutor/model_engine.py#L1236) -> From [`_get_graphs_to_capture`](https://github.com/chungen04/TensorRT-LLM/blob/e760fa2a9df60aa151c0350c64ba823d72655707/tensorrt_llm/_torch/pyexecutor/model_engine.py#L1122). 
     - Case 1: deprecating soon. 
     - Case 3: `draft_lengths = [max_total_draft_tokens]`, optionally + [0]. 
     - Case 2: `draft_len` ranges over `self._dynamic_draft_len_mapping.values()`, built by [_compute_dynamic_draft_len_mapping](https://github.com/chungen04/TensorRT-LLM/blob/e760fa2a9df60aa151c0350c64ba823d72655707/tensorrt_llm/_torch/pyexecutor/model_engine.py#L1090): for each graph batch size it picks schedule `schedule[schedule_keys[key_idx]]`, else 0. Schedule values are Pydantic-validated max_draft_len [src](https://github.com/chungen04/TensorRT-LLM/blob/e760fa2a9df60aa151c0350c64ba823d72655707/tensorrt_llm/llmapi/llm_args.py#L1131), and `max_draft_len <= max_total_draft_tokens`. `max_total_draft_tokens == max_num_draft_tokens` ([src](https://github.com/chungen04/TensorRT-LLM/blob/e760fa2a9df60aa151c0350c64ba823d72655707/tensorrt_llm/_torch/pyexecutor/py_executor_creator.py#L740), [src](https://github.com/chungen04/TensorRT-LLM/blob/e760fa2a9df60aa151c0350c64ba823d72655707/tensorrt_llm/llmapi/llm_args.py#L1202)).

File: tensorrt_llm/_torch/pyexecutor/guided_decoder.py [hunk 2]
```diff
@@ -332,9 +335,13 @@ def _apply_bitmask(self,
...
     @nvtx_range("GuidedDecoder.add_batch")
-    def add_batch(self, scheduled_requests: ScheduledRequests) -> None:
+    def add_batch(self,
+                  scheduled_requests: ScheduledRequests,
+                  runtime_draft_len: Optional[int] = None) -> None:
+        num_draft_tokens = (self.max_num_draft_tokens
+                            if runtime_draft_len is None else runtime_draft_len)
         self.requests = GuidedRequests.from_scheduled_requests(
-            scheduled_requests, self.max_num_draft_tokens)
+            scheduled_requests, num_draft_tokens)
```
[the remark this answers is not in the document]
Comment (chungen04) [reply tensorrt_llm/_torch/pyexecutor/guided_decoder.py:345]: Despite there are no explicit guarding and clamping of `runtime_draft_len`, it is currently safe that `runtime_draft_len` do not exceed `max_num_draft_tokens`, justification: 
- The caller with explicit `runtime_draft_len` assignment is at [model_engine.py](https://github.com/NVIDIA/TensorRT-LLM/pull/15023/changes#diff-203c6491ca0bedf455eeee3ed7939c3209f3f14c383594c26261e1dc48a00395R2584)
- Assignments of `runtime_draft_len` includes:
  - [ref1](https://github.com/chungen04/TensorRT-LLM/blob/e760fa2a9df60aa151c0350c64ba823d72655707/tensorrt_llm/_torch/pyexecutor/model_engine.py#L2765) -> assigned to `self.max_total_draft_tokens`.
  - [ref2 (init)](https://github.com/chungen04/TensorRT-LLM/blob/e760fa2a9df60aa151c0350c64ba823d72655707/tensorrt_llm/_torch/pyexecutor/model_engine.py#L520), [ref3 (init)](https://github.com/chungen04/TensorRT-LLM/blob/e760fa2a9df60aa151c0350c64ba823d72655707/tensorrt_llm/_torch/pyexecutor/model_engine.py#L525) -> assigned to `self.max_total_draft_tokens` or 0.
   - [ref4 (capture)](https://github.com/chungen04/TensorRT-LLM/blob/e760fa2a9df60aa151c0350c64ba823d72655707/tensorrt_llm/_torch/pyexecutor/model_engine.py#L1236) -> From [`_get_graphs_to_capture`](https://github.com/chungen04/TensorRT-LLM/blob/e760fa2a9df60aa151c0350c64ba823d72655707/tensorrt_llm/_torch/pyexecutor/model_engine.py#L1122). 
     - Case 1: deprecating soon. 
     - Case 3: `draft_lengths = [max_total_draft_tokens]`, optionally + [0]. 
     - Case 2: `draft_len` ranges over `self._dynamic_draft_len_mapping.values()`, built by [_compute_dynamic_draft_len_mapping](https://github.com/chungen04/TensorRT-LLM/blob/e760fa2a9df60aa151c0350c64ba823d72655707/tensorrt_llm/_torch/pyexecutor/model_engine.py#L1090): for each graph batch size it picks schedule `schedule[schedule_keys[key_idx]]`, else 0. Schedule values are Pydantic-validated max_draft_len [src](https://github.com/chungen04/TensorRT-LLM/blob/e760fa2a9df60aa151c0350c64ba823d72655707/tensorrt_llm/llmapi/llm_args.py#L1131), and `max_draft_len <= max_total_draft_tokens`. `max_total_draft_tokens == max_num_draft_tokens` ([src](https://github.com/chungen04/TensorRT-LLM/blob/e760fa2a9df60aa151c0350c64ba823d72655707/tensorrt_llm/_torch/pyexecutor/py_executor_creator.py#L740), [src](https://github.com/chungen04/TensorRT-LLM/blob/e760fa2a9df60aa151c0350c64ba823d72655707/tensorrt_llm/llmapi/llm_args.py#L1202)).

**Comment by chungen04** [comment]
Despite there are no explicit guarding and clamping of `runtime_draft_len`, it is currently safe that `runtime_draft_len` do not exceed `max_num_draft_tokens`, justification: 
- The caller with explicit `runtime_draft_len` assignment is at [model_engine.py](https://github.com/NVIDIA/TensorRT-LLM/pull/15023/changes#diff-203c6491ca0bedf455eeee3ed7939c3209f3f14c383594c26261e1dc48a00395R2584)
- Assignments of `runtime_draft_len` includes:
  - [ref1](https://github.com/chungen04/TensorRT-LLM/blob/e760fa2a9df60aa151c0350c64ba823d72655707/tensorrt_llm/_torch/pyexecutor/model_engine.py#L2765) -> assigned to `self.max_total_draft_tokens`.
  - [ref2 (init)](https://github.com/chungen04/TensorRT-LLM/blob/e760fa2a9df60aa151c0350c64ba823d72655707/tensorrt_llm/_torch/pyexecutor/model_engine.py#L520), [ref3 (init)](https://github.com/chungen04/TensorRT-LLM/blob/e760fa2a9df60aa151c0350c64ba823d72655707/tensorrt_llm/_torch/pyexecutor/model_engine.py#L525) -> assigned to `self.max_total_draft_tokens` or 0.
   - [ref4 (capture)](https://github.com/chungen04/TensorRT-LLM/blob/e760fa2a9df60aa151c0350c64ba823d72655707/tensorrt_llm/_torch/pyexecutor/model_engine.py#L1236) -> From [`_get_graphs_to_capture`](https://github.com/chungen04/TensorRT-LLM/blob/e760fa2a9df60aa151c0350c64ba823d72655707/tensorrt_llm/_torch/pyexecutor/model_engine.py#L1122). 
     - Case 1: deprecating soon. 
     - Case 3: `draft_lengths = [max_total_draft_tokens]`, optionally + [0]. 
     - Case 2: `draft_len` ranges over `self._dynamic_draft_len_mapping.values()`, built by [_compute_dynamic_draft_len_mapping](https://github.com/chungen04/TensorRT-LLM/blob/e760fa2a9df60aa151c0350c64ba823d72655707/tensorrt_llm/_torch/pyexecutor/model_engine.py#L1090): for each graph batch size it picks schedule `schedule[schedule_keys[key_idx]]`, else 0. Schedule values are Pydantic-validated max_draft_len [src](https://github.com/chungen04/TensorRT-LLM/blob/e760fa2a9df60aa151c0350c64ba823d72655707/tensorrt_llm/llmapi/llm_args.py#L1131), and `max_draft_len <= max_total_draft_tokens`. `max_total_draft_tokens == max_num_draft_tokens` ([src](https://github.com/chungen04/TensorRT-LLM/blob/e760fa2a9df60aa151c0350c64ba823d72655707/tensorrt_llm/_torch/pyexecutor/py_executor_creator.py#L740), [src](https://github.com/chungen04/TensorRT-LLM/blob/e760fa2a9df60aa151c0350c64ba823d72655707/tensorrt_llm/llmapi/llm_args.py#L1202)).

**Comment by achartier** [comment]
Could you clarify the intended order between this PR and https://github.com/NVIDIA/TensorRT-LLM/pull/15026?

**Comment by chungen04** [comment]
@achartier I am not sure about the content and author in #15026 , looks like he proposed a PR quickly soon after my issue & PR, but there are some design choices to be justified there.

File: tensorrt_llm/_torch/pyexecutor/guided_decoder.py [hunk 3]
```diff
@@ -332,9 +335,13 @@ def _apply_bitmask(self,
             d2t=d2t)
 
     @nvtx_range("GuidedDecoder.add_batch")
-    def add_batch(self, scheduled_requests: ScheduledRequests) -> None:
+    def add_batch(self,
+                  scheduled_requests: ScheduledRequests,
+                  runtime_draft_len: Optional[int] = None) -> None:
```
Comment (achartier) [inline tensorrt_llm/_torch/pyexecutor/guided_decoder.py:341]: Which caller site was updated to set `runtime_draft_len`? Is one file missing from the review?

Comment (chungen04) [reply tensorrt_llm/_torch/pyexecutor/guided_decoder.py:341]: It is [here](https://github.com/NVIDIA/TensorRT-LLM/pull/15023/changes#diff-203c6491ca0bedf455eeee3ed7939c3209f3f14c383594c26261e1dc48a00395R2584), L2584 in model_engine.py. See another file requested for review.

Comment (achartier) [reply tensorrt_llm/_torch/pyexecutor/guided_decoder.py:341]: I think it's always a `CapturableGuidedDecoder` at line 2584, but it's fine to keep the APIs of the 2 classes synchronized.

**Comment by achartier** [comment]
Please follow these steps to setup pre-commit and run it on this PR: https://github.com/NVIDIA/TensorRT-LLM/blob/main/CONTRIBUTING.md

**Comment by chungen04** [comment]
@achartier Done, please check

**Comment by achartier** [comment]
@chungen04 Could you double check? `yapf` is still failing on the last version.

**Comment by chungen04** [comment]
@achartier Done, please check. Looks like I missed a message saying yapf was installing on my end.

**Review by zheyuf — APPROVED** [review]
Looks good to me for the related draft_len_schedule change. Thanks for the PR!

**Review by achartier — APPROVED** [review]

**Comment by chungen04** [comment]
@achartier Should I do anything regarding the CI failure?

**Comment by achartier** [comment]
> @achartier Should I do anything regarding the CI failure?

Infra issue. I rebased and relaunched the pipeline.

**Comment by chungen04** [comment]
Oh, here we go. Thank you @achartier for the support!
